storage/innobase/row/row0ins.cc

```diff
@@ -2711,21 +2711,44 @@ row_ins_clust_index_entry_low(
 
 	DBUG_EXECUTE_IF("row_ins_row_level", goto skip_bulk_insert;);
 
-	if (!(flags & BTR_NO_UNDO_LOG_FLAG)
-	    && page_is_empty(block->page.frame)
-	    && !entry->is_metadata() && !trx->duplicates
-	    && !trx->check_unique_secondary && !trx->check_foreigns
-	    && !trx->dict_operation
+	if (srv_allow_bulk_insert
+            && ((!(flags & BTR_NO_UNDO_LOG_FLAG)
+	         && page_is_empty(block->page.frame)
+	         && !entry->is_metadata() && !trx->duplicates
+	         && !trx->check_unique_secondary && !trx->check_foreigns
+	         && !trx->dict_operation
+	         && !index->table->skip_alter_undo
+	         && !index->table->is_active_ddl())
+	        || (index->table->skip_alter_undo
+		    && !index->table->is_temporary()
+		    && index->table->foreign_set.empty()))
 	    && block->page.id().page_no() == index->page
-	    && !index->table->skip_alter_undo
 	    && !index->table->n_rec_locks
-	    && !index->table->is_active_ddl()
 	    && !index->table->has_spatial_index()
 	    && !index->table->versioned()) {
 		DEBUG_SYNC_C("empty_root_page_insert");
 
-		trx->bulk_insert = true;
+		/* Ignore the alter bulk insert operation for
+		temporary or versioning table.
+		In case of foreign key constraint,
+		InnoDB has to check the constraint validity for
+		each row. But bulk insert operation doesn't
+		check foreign constraint. Skip bulk insert
+		operation if the table contains foreign key
+		relation */
+		if (index->table->skip_alter_undo) {
+			trx_start_if_not_started(trx, true);
+			trx->bulk_insert= true;
```

Comment:
I do not see any code that refers to a `FOREIGN KEY` here. I suppose that it is being checked at a higher level.

It could be useful to add a comment that we are only executing this code at the start of an `ALTER TABLE` statement if that is the case, and note how and where subsequent written rows will get processed.

Maybe the new Boolean configuration parameter and its name should be specific to only this code path, for example as follows:
```c++
if (UNIV_LIKELY(!index->table->skip_alter_undo)) {
	trx->bulk_insert = true;
} else if (!innodb_alter_copy_bulk) {
	goto skip_bulk_insert;
} else {
	/* We are writing the first row in ALTER TABLE. */
```